draco/analysis/hyforesbandpass.py

```diff
@@ -0,0 +1,520 @@
+                    tvar = tools.invert_no_zero(weight[pp, :, xx, tt]) ### inverse of the weight is variance (keep zeros zeros). Looks like the variance is taken over DEC or el.
+
+                    # Grab the filter for this pol and ew baseline
+                    NF = np.ascontiguousarray(filt[pp, :, :, xx, tt])
+
+                    # Make sure that any frequencies unmasked during filter generation
+                    # are also unmasked in the data
+                    valid_freq_flag = np.any(np.abs(NF) > 0.0, axis=0)
+
+                    if not np.any(valid_freq_flag):   ### if the filter is entirely zero --> skip (cond 2/3)
+                        # Skip samples where filter is entirely zero 
+                        weight[pp, :, xx, tt] = 0.0
+                        continue
+
+                    missing_freq = np.flatnonzero(valid_freq_flag & ~flag) ### find the indices of frequencies needed by the delay filter but missing from the unfiltered data --> skip (cond 3/3)
+                    if missing_freq.size > 0:
+                        self.log.warning(
+                            "Missing the following frequencies that were "
+                            "assumed valid during filter generation: "
+                            f"{missing_freq}"
+                        )
+                        weight[pp, :, xx, tt] = 0.0
+                        continue
+                    
+                    # Implement Step 5 
+                    diag_m = 1 - g[pp, xx, :]
+                    vis[pp, :, xx, :, tt] = np.matmul(NF, diag_m[:, np.newaxis]*tvis) ### compute using contiguous filter and visibility data ### this must have changed hv.vis
+                    weight[pp, :, xx, tt] = tools.invert_no_zero( ### variance multiples with the filter value squared
+                        np.matmul(np.abs(NF) ** 2, (np.abs(diag_m[:])**2)*tvar))
+                    # Flag frequencies with large attenuation ### mask frequencies where in the filtered data they do not have enough same-frequency contributions from the unfiltered data
+                    if self.atten_threshold > 0.0:
+                        diag = np.abs(np.diag(NF))
+                        nonzero_diag_flag = diag > 0.0
+                        if np.any(nonzero_diag_flag):
+                            med_diag = np.median(diag[nonzero_diag_flag])
+                            flag_low = diag > (self.atten_threshold * med_diag)
+                            weight[pp, :, xx, tt] *= flag_low.astype(weight.dtype) ### this masking is only done on the weight
+
+            self.log.debug(f"Took {time.time() - t0:0.3f} seconds in total.")
+
+        # Problems saving to disk when distributed over last axis
+        hv.redistribute("freq") ### why redistribute again? How does this work when the data are aready distributed over the ra axis on different nodes?
```

Comment:
Typically we keep datasets distributed over frequency because most operations we do are frequency-independent, and it's the fastest axis to distribute over when writing files to disk.